
go_folder = "5_gene_ontology"

rule download_go_associations:
    params:
        obo="http://purl.obolibrary.org/obo/go.obo",
        gene2go="https://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2go.gz",
        gene2ensembl="https://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2ensembl.gz"
    conda: "../environments/goea.yml"
    output:
        obo=os.path.join(OUTPUTDIR, go_folder, "data/go-basic.obo"), # contains graph of go-term associations
        gene2go=os.path.join(OUTPUTDIR, go_folder, "data/gene2go"), # contains gene to go-term mapping
        gene2ensembl=os.path.join(OUTPUTDIR, go_folder, "data/gene2ensembl") # contains Entrez GeneID to Ensembl gene-id mapping
    script:
        "../scripts/download_go_associations.py"

# Get all protein coding genes as background for GO enrichment analysis
rule download_background_genes:
    input:
        gtf=config['annotation']['gtf'],
        gene2ensembl=rules.download_go_associations.output.gene2ensembl
    params:
        email="hendrik.schultheis@mpi-bn.mpg.de" # TODO add to config
    conda: "../environments/goea.yml"
    output:
        genes=os.path.join(OUTPUTDIR, go_folder, "data/protein_coding_genes.tsv")
    script:
        "../scripts/download_background_genes.py"

# rule go_enrichment_analysis:
#     input: 
#         obo=rules.download_go_associations.output.obo,
#         gene2go=rules.download_go_associations.output.gene2go,
#         #gene2ensembl=rules.download_go_associations.output.gene2ensembl,
#         annotation=rules.uropa_annotate_motifs.output.dir#,
#         #bg_genes=rules.download_background_genes.output.genes
#     params:
#         email="hendrik.schultheis@mpi-bn.mpg.de"
#     conda: "../environments/goea.yml"
#     output:
#         dir=directory(os.path.join(OUTPUTDIR, go_folder, {motif}))
#     script:
#         "../scripts/go_enrichment_analysis.py"

# rule plot_go_tree:
#     input:
#     output:
#     script: